src/daft-logical-plan/src/optimization/rules/granular_projections.rs

```diff
@@ -0,0 +1,389 @@
+                            changed = true;
+                            // Split and save child expression
+                            // Child may not have an alias, so we need to generate a new one
+                            // TODO: Remove with ordinals
+                            let child_name = format!("id-{}", uuid::Uuid::new_v4());
+
+                            let child = child.alias(child_name.as_str());
+                            split_exprs.push(child);
+                            new_children[idx] = Arc::new(Expr::Column(daft_dsl::Column::Resolved(
+                                daft_dsl::ResolvedColumn::Basic(child_name.into()),
+                            )));
+                        }
+                    }
+
+                    if changed {
+                        let expr = e.with_new_children(new_children);
+                        return Ok(Transformed::yes(Arc::new(expr)));
+                    } else {
+                        return Ok(Transformed::no(e));
+                    }
+                }
+
+                Ok(Transformed::no(e))
+            })?;
+
+            // Push the top level expression that was changed
+            split_exprs.push(res.data);
+            all_split_exprs.push(split_exprs);
+        }
+
+        // In this case, none of the projection columns were split or changed, so we can just continue
+        if all_split_exprs.iter().all(|x| x.len() == 1) {
+            return Ok(Transformed::no(plan));
+        }
+
+        // Create new Project nodes. There will be as many Project nodes as there are
+        // maximum # of split expressions for one of the original projection columns
+        let max_projections = all_split_exprs.iter().map(|x| x.len()).max().unwrap_or(1);
+
+        // Construct new Project nodes in the following way:
+        // Original: Project(input, [e1, e2, e3])
+        // Split: [e1, e2, e3] -> [[e11, e12, e13], [e21], [e31, e32]]
+        // New:
+        //  - Project(input, [e11, e21, e31])
+        //  - Project(input, [e12, passthrough, e32])
+        //  - Project(input, [e13, passthrough, passthrough])
+        // Passthrough means just passing the original column through unchanged
+
+        let mut last_child = input.clone();
+        for i in 0..max_projections {
+            let exprs = all_split_exprs
+                .iter()
+                .map(|split| {
+                    split.get(i).cloned().unwrap_or_else(|| {
+                        let passthrough_name = split.last().unwrap().name();
+
+                        Arc::new(Expr::Column(daft_dsl::Column::Resolved(
+                            daft_dsl::ResolvedColumn::Basic(passthrough_name.into()),
+                        )))
+                    })
+                })
+                .collect_vec();
+
+            last_child = Arc::new(LogicalPlan::Project(
+                Project::try_new(last_child, exprs).unwrap(),
+            ));
+        }
```

Comment:
I think I understand what you are doing here. I don't have any complaints about it, but a suggestion for an alternative design is to do a top-down traversal of the plan and expression trees, and only split one layer of expressions per iteration. Then, if there are children expression that also need to be split, they will be handled in the subsequent traversal, since it would go into the newly produced child projection.